### PR TITLE
Massive Set of Changes

### PR DESCRIPTION
**Pull request by larryk85**
I apologize in advance for the review.

For playing around with this look at https://github.com/larryk85/antler-proj-foo, this should have a reasonably complex project that you can build and populate.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,25 +1,32 @@
...
-cmake_minimum_required(VERSION 3.11) # If ryml >= 0.5.0 is included via FetchContent or submodule, this minimum is 3.12.
+cmake_minimum_required(VERSION 3.10)
 
 project("ANTLER Project Tools " VERSION 1.0.0)
 
 include( ./common.cmake REQUIRED )
 
+
+
+option(ANTLER_PROJ_PEDANTIC_BUILD "Pedantic builds" On)
```
Comment (ScottBailey) [inline CMakeLists.txt:11]: Is this used anywhere?

File: include/antler/project/dependency.hpp [hunk 2]
```diff
@@ -39,21 +46,41 @@ class dependency {
...
+      // we only parsed a name
+      if (proj.size() == s.size()) {
+         m_name = std::move(proj);
+      } else {
+         ANTLER_CHECK(proj.size() + nm.size() == s.size()-1, "Invalid dependency name: {}", s );
+         m_proj = std::move(proj);
+         m_name = std::move(nm); 
+      }
+   }
+   */
```
Comment (ScottBailey) [inline include/antler/project/dependency.hpp:76]: This looks like some iterative dev code that accidentally got committed? 
```suggestion
```

File: include/antler/project/dependency.hpp [hunk 3]
```diff
@@ -39,21 +46,41 @@ class dependency {
...
+   void set(std::string name, std::string_view loc, std::string_view tag, std::string_view rel, std::string_view hash);
 
    /// Get the dependency name.
    /// @return The name of this dependency.
-   [[nodiscard]] std::string_view name() const noexcept;
+   [[nodiscard]] inline const std::string& name() const noexcept { return m_name; }
+
+   ///// Get the project name of the dependency.
+   ///// @return The name of the project.
+   //[[nodiscard]] inline const std::string& proj() const noexcept { return m_proj; }
```
Comment (dimas1185) [inline include/antler/project/dependency.hpp:57]: please remove unnecessary commented out code unless it was done on purpose

File: include/antler/project/dependency.hpp [hunk 4]
```diff
@@ -65,17 +92,18 @@ class dependency {
...
-   [[nodiscard]] std::string_view tag() const noexcept;
+   [[nodiscard]] const std::string& tag() const noexcept;
    /// Set github tag/commit hash.
    /// @param s  The new github tag/commit hash.
    void tag(std::string_view s) noexcept;
 
    /// Get the github release version.
    /// @return The possibly empty release version.
-   [[nodiscard]] std::string_view release() const noexcept;
+   [[nodiscard]] std::string release() const noexcept { return m_rel; }
```
Comment (dimas1185) [inline include/antler/project/dependency.hpp:102]: this file have mixed types, sometimes return value is `const std::string&`, sometimes `std::string` and sometimes `std::string_view`. It would be nice to make it consistent and to use only one unless there is reason for exception

File: include/antler/project/mustache.hpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+      std::vector<kainjow::mustache::data> internal_list;
+   };
+
+   template <typename T>
+   struct mustache_list<T> {
+      mustache_list()
+   };
+
+} // namespace antler::project
```
Comment (dimas1185) [inline include/antler/project/mustache.hpp:42]: I do not see any of those classes are used by simple text search. are they?

File: include/antler/project/project.hpp [hunk 6]
```diff
@@ -49,132 +63,162 @@ class project {
...
+   inline void upsert(object<Tag>&& obj) noexcept {
+      if constexpr (std::is_same_v<Tag, app_tag>)
          return upsert_app(std::move(obj));
-      else if constexpr (Ty == object::type_t::lib)
+      else if constexpr (std::is_same_v<Tag, lib_tag>)
          return upsert_lib(std::move(obj));
-      else if constexpr (Ty == object::type_t::test)
-         return upsert_test(std::move(obj));
+      //else if constexpr (Ty == object::type_t::test)
+      //   return upsert_test(std::move(obj));
```
Comment (dimas1185) [inline include/antler/project/project.hpp:91]: any reason for keeping commented out code?

File: tests/location_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <catch2/catch.hpp>
+
+TEST_CASE("Testing location clone") {
+   using namespace antler::project;
+
+   antler::system::fs::remove_all("./foo");
+
+   location::clone_github_repo("antelopeio", "antler-proj", "main", 10, "./foo/foo2");
+   CHECK(antler::system::fs::exists("./foo/foo2/.git"));
```
Comment (dimas1185) [inline tests/location_tests.cpp:13]: antler-proj/tests/location_tests.cpp:13: FAILED:
  CHECK( antler::system::fs::exists("./foo/foo2/.git") )
with expansion:
  false

File: tests/location_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+   //location::clone_git_repo("https://github.com/larryk85/cturtle", "main", 10, "./foo/foo3");
+   //CHECK(antler::system::fs::exists("./foo/foo3/.git"));
```
Comment (dimas1185) [inline tests/location_tests.cpp:16]: please remove if commented out code is unnecessary

File: tests/string_from.cpp [hunk 9]
```diff
@@ -15,6 +15,8 @@ bool same(const L& l, const R& r) {
 
 using namespace antler::string;
 
+TEST_CASE("Testing antler::string")
```
Comment (dimas1185) [inline tests/string_from.cpp:18]: unused test for deleted code. please remove

File: tools/CMakeLists.txt [hunk 10]
```diff
@@ -2,4 +2,16 @@
...
+
+macro(add_cli_test name file)
+   configure_file(${CMAKE_CURRENT_SOURCE_DIR}/tests/${file} ${CMAKE_CURRENT_BINARY_DIR}/tests/${file} COPYONLY)
+   add_test(NAME ${name} COMMAND python3 ${file} WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/tests)
+endmacro()
+
+if (ANTLER_PROJ_BUILD_TESTS)
+   enable_testing()
+   configure_file(${CMAKE_CURRENT_SOURCE_DIR}/tests/util.py ${CMAKE_CURRENT_BINARY_DIR}/tests/util.py COPYONLY)
+   #add_cli_test(add_to_project_tests add_to_tests.py)
```
Comment (dimas1185) [inline tools/CMakeLists.txt:15]: is this code for future?

File: tools/build.hpp [hunk 11]
```diff
@@ -15,92 +14,87 @@ namespace antler {
...
-
-         if (!std::filesystem::exists(build)) {
-            return true;
-         }
-
-         auto last_pop_time      = std::filesystem::last_write_time(build);
-
-         return last_pop_time < last_manifest_time;
+         subcommand->add_option("-p, path", path, "This is the path to the root of the project.")->default_val(".");
+         subcommand->add_option("-j, --jobs", jobs, "This is the path to the root of the project.")->default_val(1);
```
Comment (dimas1185) [inline tools/build.hpp:18]: wrong description

File: tools/main.cpp [hunk 12]
```diff
@@ -66,5 +65,9 @@ int main(int argc, char** argv) {
 
    CLI11_PARSE(app, argc, argv);
 
-   return runner.exec();
+   try {
+      return runner.exec();
+   } catch(...) {
+      return -1;
```
Comment (dimas1185) [inline tools/main.cpp:71]: it would be nice to have more verbose exceptions handling unless runner is not supposed to print all the details

File: tools/remove_from.hpp [hunk 13]
```diff
@@ -12,63 +12,60 @@
...
          }
          return false;
       }
 
-      inline bool remove_app(antler::project::project& proj) { return remove_obj<antler::project::object::type_t::app>(proj); }
-      inline bool remove_lib(antler::project::project& proj) { return remove_obj<antler::project::object::type_t::lib>(proj); }
-      inline bool remove_test(antler::project::project& proj) { return remove_obj<antler::project::object::type_t::test>(proj); }
+      inline bool remove_app(antler::project::project& proj) { return remove_obj<antler::project::app_t>(proj); }
+      inline bool remove_lib(antler::project::project& proj) { return remove_obj<antler::project::lib_t>(proj); }
+      //inline bool remove_test(antler::project::project& proj) { return remove_obj<antler::project::test_t>(proj); }
```
Comment (dimas1185) [inline tools/remove_from.hpp:31]: my understanding is that test is not working yet so you commented this out? it would be nice to have TODO section to not mess this with debug or unnecessary code

File: tools/remove_from.hpp [hunk 14]
```diff
@@ -12,63 +12,60 @@
...
-                     std::cerr << "Dependency: " << dep_name << " is not a dependency of: " << obj_name << std::endl;
-                     return false;
-                  }
-               } catch(...) {
-                  std::cerr << "Object: " << obj_name << " does not exist." << std::endl;
+      bool remove_dependency_from_all(antler::project::project& proj) {
+         const auto& remove_dep = [](auto dep, auto& objs) {
+            for (auto& [k, o] : objs) {
+               if (o.remove_dependency(dep)) {
+                  std::cout << "Removing dependency: " << dep << " from: " << o.name() << std::endl;
```
Comment (dimas1185) [inline tools/remove_from.hpp:37]: I assume system::info_log supposed to be used here

File: tools/tests/init_tests.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+from util import *
+
+def init_project(project_name, version, path):
+   out, err = antler_proj_cmd("init -p {0} -n {1} -v {2}".format(path, project_name, version))
+   print(out)
+
+
+def test_init_project():
+   shutil.rmtree("test0")
```
Comment (dimas1185) [inline tools/tests/init_tests.py:13]: File ".../antler-proj/build/tools/tests/init_tests.py", line 41, in <module>
    test_init_project()
  File ".../antler-proj/build/tools/tests/init_tests.py", line 13, in test_init_project
    shutil.rmtree("test0")
  File "/usr/lib/python3.10/shutil.py", line 714, in rmtree
    onerror(os.lstat, path, sys.exc_info())
  File "/usr/lib/python3.10/shutil.py", line 712, in rmtree
    orig_st = os.lstat(path)
FileNotFoundError: [Errno 2] No such file or directory: 'test0'

**Review by dimas1185 — CHANGES_REQUESTED** [review]
order of importance:
* 2 tests are failing: `init_tests` `Testing location clone`
* other stuff is cleanup & refactoring suggestions for about 15-30 minutes

**Comment by jolly-fellow** [comment]
I have tested this new update. It fixed a couple of bugs but added two new, all other bugs are still actual. Please take a look at https://github.com/AntelopeIO/antler-proj/issues I have described them there.

File: CMakeLists.txt [hunk 16]
```diff
@@ -1,33 +1,40 @@
...
+  find_program(CLANG_TIDY_EXE NAMES "clang-tidy" REQUIRED)
+  set(CLANG_TIDY_COMMAND "${CLANG_TIDY_EXE}" "-checks=*,modernize-*")
+endif()
 
 # Copy licenses to BUILD dir
 #
 #  NOTE: This is used due to legacy requirements instead of:
 #    `install(FILE <license files> COMPONENT Documentation DESTINATION share/aproj/licenses)`
-configure_file(LICENSE ${CMAKE_BINARY_DIR}/licenses/LICENSE COPYONLY)
+configure_file(LICENSE ${CMAKE_CURRENT_BINARY_DIR}/licenses/LICENSE COPYONLY)
```
Comment (ScottBailey) [inline CMakeLists.txt:35]: I am beginning to think we should create a `copy_license()` function. But in the meantime, I suggest we make these more readable. 
```suggestion
configure_file(LICENSE ./licenses/ COPYONLY)
```

File: CMakeLists.txt [hunk 17]
```diff
@@ -1,33 +1,40 @@
...
 file(GLOB license_files licenses/*)
 foreach(full_path IN ITEMS ${license_files})
-  string(REPLACE "${CMAKE_SOURCE_DIR}/" "" file_name ${full_path})
-  configure_file("${file_name}" "${CMAKE_BINARY_DIR}/${file_name}" COPYONLY)
+	string(REPLACE "${CMAKE_SOURCE_DIR}/" "" file_name ${full_path})
+	#configure_file("${file_name}" "${CMAKE_CURRENT_BINARY_DIR}/${file_name}" COPYONLY)
```
Comment (ScottBailey) [inline CMakeLists.txt:39]: If it's commented out, we should document why and what the criteria for enabling is. But I'm wondering if it shouldn't just be like so?
```suggestion
	configure_file("${file_name}" ./licenses/ COPYONLY)
```

File: include/antler/project/location.hpp [hunk 18]
```diff
@@ -22,12 +25,18 @@ namespace antler::project::location {
...
 bool is_reachable(std::string_view l);
 /// @param l  Location to evaluate.
 /// @return true if l looks like a url.
 bool is_url(std::string_view l);
 
+bool clone_github_repo(const std::string& org, const std::string& repo, const std::string& branch, uint32_t jobs, system::fs::path dest = ".");
+bool clone_git_repo(const std::string& url, const std::string& branch, uint32_t jobs, system::fs::path dest = ".");
+bool pull_git_repo(system::fs::path src);
+std::string github_request(const std::string& org, const std::string& repo);
+std::string get_github_default_branch(const std::string& org, const std::string& repo);
```
Comment (ScottBailey) [inline include/antler/project/location.hpp:40]: Shouldn't these be marked `nodiscard`? 
```suggestion
[[nodiscard]] bool clone_github_repo(const std::string& org, const std::string& repo, const std::string& branch, uint32_t jobs, system::fs::path dest = ".");
[[nodiscard]] bool clone_git_repo(const std::string& url, const std::string& branch, uint32_t jobs, system::fs::path dest = ".");
[[nodiscard]] bool pull_git_repo(system::fs::path src);
[[nodiscard]] std::string github_request(const std::string& org, const std::string& repo);
[[nodiscard]] std::string get_github_default_branch(const std::string& org, const std::string& repo);
```

File: include/antler/project/location.hpp [hunk 19]
```diff
@@ -22,12 +25,18 @@ namespace antler::project::location {
...
+[[nodiscard]] bool is_github_shorthand(std::string_view l);
 /// @param l  Location to evaluate.
 /// @return true if l is reachable.
 bool is_reachable(std::string_view l);
 /// @param l  Location to evaluate.
 /// @return true if l looks like a url.
 bool is_url(std::string_view l);
 
+bool clone_github_repo(const std::string& org, const std::string& repo, const std::string& branch, uint32_t jobs, system::fs::path dest = ".");
+bool clone_git_repo(const std::string& url, const std::string& branch, uint32_t jobs, system::fs::path dest = ".");
```
Comment (ScottBailey) [inline include/antler/project/location.hpp:37]: I'm thinking `jobs` should become the final parameter and default to `0`  OR maybe it's simpler and better to honor the users `submodule.fetchJobs` setting.

File: include/antler/project/net_utils.hpp [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+      inline bool is_reachable(std::string_view url) {
+         try {
+            (void)request(url);
+            return true;
+         } catch(...) {
+            return false;
+         }
+      }
+
+      CURL* curl_obj;
```
Comment (ScottBailey) [inline include/antler/project/net_utils.hpp:80]: Leaving this public feels like a latent defect waiting to happen.
```suggestion
private:
      CURL* curl_obj;
```

File: include/antler/project/net_utils.hpp [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+      }
+
+      static inline bool is_shorthand(std::string_view s) {
+         auto sub = s.substr(0, s.find_last_of("/"));
+         return sub.size() != s.size() && sub.find_last_of("/") == std::string_view::npos;
+      }
+
+      curl sender;
+      std::string bearer_token = "";
+      bool has_gh_app          = false;
```
Comment (ScottBailey) [inline include/antler/project/net_utils.hpp:152]: I don't believe these should be used outside of the scope of this struct/class so it makes sense to protect them:
```suggestion
   private:
      curl sender;
      std::string bearer_token = "";
      bool has_gh_app          = false;
```

File: include/antler/project/net_utils.hpp
Comment (ScottBailey) [inline include/antler/project/net_utils.hpp:1]: Suggest most functions receive `nodiscard` decoration, NBD if not.

File: include/antler/project/populator.hpp [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+               inst.m_populators.emplace(proj.name(), populator(proj));
+            }
+            return inst.m_populators.at(proj.name());
+         }
+
+         static populator& get(std::string_view name) {
+            return instance().m_populators.at(name);
+         }
+
+         static void add_mapping(std::string l, std::string mapping) {
```
Comment (ScottBailey) [inline include/antler/project/populator.hpp:69]: Please make first parameter (`l`) more descriptive or document.

File: include/antler/project/project.hpp [hunk 23]
```diff
@@ -49,132 +63,162 @@ class project {
...
 
    /// Return the first object with the matching name where search order is apps, libs, tests.
    /// @TODO replace this with a std::vector<antler::project::object>/antler::project::object::list_t to return all the objects
    /// with matching names.
    /// @param name  The name to search for in the object lists.
    /// @param type  If type is other than any, the search is limited to that single type.
    /// @return vector with copies of the objects.
-   [[nodiscard]] antler::project::object& object(std::string_view name); 
+   template <typename Tag>
+   [[nodiscard]] auto& object(const std::string& name) {
```
Comment (ScottBailey) [inline include/antler/project/project.hpp:136]: Is the doxygen comment for this still correct, I think it's not.

Comment (ScottBailey) [inline include/antler/project/project.hpp:136]: RN, the doxygen comment references tests, but the code doesn't follow. Yet. Plans?

File: include/antler/project/project.hpp [hunk 24]
```diff
@@ -49,132 +63,162 @@ class project {
...
+   }
 
-private:
-   inline void create_app(std::string_view name) {
+   /// Deserialization function from yaml node to version
+   [[nodiscard]] inline bool from_yaml(const yaml::node_t& n) noexcept {
+      return ANTLER_EXPECT_YAML(n, "project", name, std::string) &&
+             ANTLER_EXPECT_YAML(n, "version", version, antler::project::version) &&
+             ANTLER_TRY_YAML_ALL(n, "apps", upsert, antler::project::app_t) &&
+             ANTLER_TRY_YAML_ALL(n, "libs", upsert, antler::project::lib_t);
```
Comment (ScottBailey) [inline include/antler/project/project.hpp:207]: This is short-circuit code and the logger will only report the first missing label. Would it be better to do this as a series of `&=` calls?

File: include/antler/project/yaml.hpp [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+namespace antler::project {
+   struct yaml {
+      using node_t = YAML::Node;
+
+      inline static node_t load(const system::fs::path& path) { return YAML::LoadFile(path.string()); }
+
+      inline static void write(system::fs::path path, const node_t& node) {
+         std::ofstream out(path);
+         out << node;
+         out.close();
```
Comment (ScottBailey) [inline include/antler/project/yaml.hpp:26]: If we aren't checking for or logging errors, why are we manually `closing()`ing instead of the destructor?

File: include/antler/system/utils.hpp [hunk 26]
```diff
@@ -72,8 +72,24 @@ namespace antler::system {
       std::size_t n;
 
       while ((n = fread(buff.data(), 1, buff.size(), h)) > 0) {
-         fwrite(buff.data(), 1, n, stdout);
+         fwrite(buff.data(), n, 1, stdout);
```
Comment (ScottBailey) [inline include/antler/system/utils.hpp:75]: Why the change? In the read, we are reading `n` chars of data, in the write we are writing one block of data with length `n`?
```
size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream);
size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream);
   ...'nmemb' items of data, each 'size' bytes long...
```

```suggestion
         fwrite(buff.data(), 1, n, stdout);
```

File: include/antler/system/utils.hpp [hunk 27]
```diff
@@ -47,23 +55,15 @@ namespace antler::system {
...
-      auto sv = magic_enum::enum_cast<Enum>(s);
-      if (sv) {
-         return *sv;
-      } else {
-         return Enum::error;
+   inline static int32_t execute(std::string cmd, const std::vector<std::string>& args) {
+      std::string_view prog = cmd;
+      for (const auto& arg : args) {
+         cmd += " " + arg;
       }
```
Comment (ScottBailey) [inline include/antler/system/utils.hpp:62]: This needs to be fixed: prog may not be valid if cmd is resized. 
```suggestion
   inline static int32_t execute(std::string_view prog, const std::vector<std::string>& args) {
      std::string cmd{prog};
      for (const auto& arg : args) {
         cmd += " " + arg;
      }
```

File: include/antler/system/utils.hpp [hunk 28]
```diff
@@ -72,8 +72,24 @@ namespace antler::system {
...
       while ((n = fread(buff.data(), 1, buff.size(), h)) > 0) {
-         fwrite(buff.data(), 1, n, stdout);
+         fwrite(buff.data(), n, 1, stdout);
       }
+
+      return WEXITSTATUS(pclose(h));
+   }
+
+   inline static int32_t execute_quiet(std::string cmd, const std::vector<std::string>& args) {
+      std::string_view prog = cmd;
```
Comment (ScottBailey) [inline include/antler/system/utils.hpp:82]: Same defect as above.
```suggestion
   inline static int32_t execute_quiet(std::string_view prog, const std::vector<std::string>& args) {
      std::string cmd{prog};
```

File: include/antler/system/utils.hpp [hunk 29]
```diff
@@ -47,23 +55,15 @@ namespace antler::system {
...
-
-   template <typename Enum, typename = std::enable_if_t<std::is_enum_v<Enum>>>
-   inline static Enum from_string(std::string_view s) { 
-      auto sv = magic_enum::enum_cast<Enum>(s);
-      if (sv) {
-         return *sv;
-      } else {
-         return Enum::error;
+   inline static int32_t execute(std::string cmd, const std::vector<std::string>& args) {
+      std::string_view prog = cmd;
```
Comment (ScottBailey) [inline include/antler/system/utils.hpp:59]: This needs to be fixed: prog may not be valid if cmd is resized, which may happen if args is populated.
```suggestion
   inline static int32_t execute(std::string_view prog, const std::vector<std::string>& args) {
      std::string cmd{prog};
```

File: licenses/LICENSE.Mustache
Comment (ScottBailey) [inline licenses/LICENSE.Mustache:1]: I think this should be an MIT license?
https://github.com/no1msd/mstch/blob/master/LICENSE

File: src/dependency.cpp [hunk 30]
```diff
@@ -28,10 +28,6 @@ inline bool is_valid_hash(std::string_view s, size_t byte_count = 32) noexcept {
 namespace antler::project {
 
 
-//--- constructors/destructor ------------------------------------------------------------------------------------------
-
-
-
 //--- alphabetic --------------------------------------------------------------------------------------------------------
```
Comment (ScottBailey) [inline src/dependency.cpp:31]: Might want to remove this comment or alphabetize the functions.

File: src/location.cpp [hunk 31]
```diff
@@ -150,12 +28,19 @@ bool is_github_archive(std::string_view s) { return is_github(s) && is_archive(s
...
 
-bool is_github_org_repo_shorthand(std::string_view s) { return github::is_shorthand(s); }
+bool is_github_shorthand(std::string_view s) { return github::is_shorthand(s); }
 
 bool is_github_repo(std::string_view s) { return is_github(s) && !is_archive(s); }
 
 bool is_reachable(std::string_view l) {
-   if (is_github_repo(l) || is_github_org_repo_shorthand(l)) {
+   if (!is_github_shorthand(l)) {
+      system::error_log("In this version of antler-proj only github shorthands are supported. Generalized git repos and archives will be supported in the next version.");
```
Comment (ScottBailey) [inline src/location.cpp:37]: Maybe we want to under promise and over deliver here?
```suggestion
      system::error_log("In this version of antler-proj only github shorthands are supported. Generalized git repos and archives will be supported in a future version.");
```

File: tests/init_tests.cpp
Comment (ScottBailey) [inline tests/init_tests.cpp:1]: Test failure:
image

File: tools/build.hpp (hunk 11 above)
Comment (ScottBailey) [reply tools/build.hpp:18]: I don't think `1` is a good default value for jobs. An acceptable value might be `0` or the user's `submodule.fetchJobs` setting. It doesn't seem friendly to override the users gitconfig, though.

File: tools/remove_from.hpp [hunk 32]
```diff
@@ -12,63 +12,60 @@
...
+
+         if (obj.remove_dependency(dep_name)) {
+            system::info_log("Dependency {0} removed from object {1}.", dep_name, obj_name);
+            return true;
+         } else {
+            system::error_log("Dependency {0} is not a dependency of object {1}.", dep_name, obj_name);
+            return false;
+         }
+         
          return true;
```
Comment (ScottBailey) [inline tools/remove_from.hpp:62]: Can't get here. Looks funny.
```suggestion
```

**Comment by ScottBailey** [comment]
The following files have "wrong" line endings:
```
include/antler/project/source.hpp
include/antler/system/utils.hpp
include/antler/system/version.hpp.in
tests/cmake_tests.cpp
tests/init_tests.cpp
tests/main.cpp
tools/add_to.hpp
tools/app_version.hpp
tools/build.hpp
tools/common.hpp
tools/init.hpp
tools/main.cpp
tools/populate.hpp
tools/remove_from.hpp
tools/update.hpp
tools/validate.hpp
```

**Review by ScottBailey — CHANGES_REQUESTED** [review]
See comments.

Must fix 1) string_views pointing to strings that are likely to resize; and 2) correct or skip tests.

File: CMakeLists.txt (hunk 16 above)
Comment (ScottBailey) [reply CMakeLists.txt:35]: see 1afe6e5f11b6f879d182ae792838066e6e3f96e7

File: CMakeLists.txt (hunk 17 above)
Comment (ScottBailey) [reply CMakeLists.txt]: See commit 1afe6e5

Comment (ScottBailey) [reply CMakeLists.txt]: removed in 3874966, can be reinserted if/when needed.

File: include/antler/project/location.hpp (hunk 18 above)
Comment (ScottBailey) [reply include/antler/project/location.hpp]: Fixed in 1e42809

File: tools/build.hpp (hunk 11 above)
Comment (ScottBailey) [reply tools/build.hpp]: Description corrected in 55f8937.
#28 written to address unfriendly jobs behavior.

File: include/antler/project/location.hpp (hunk 19 above)
Comment (ScottBailey) [reply include/antler/project/location.hpp]: #28 written to address unfriendly jobs behavior.

File: CMakeLists.txt (hunk 1 above)
Comment (ScottBailey) [reply CMakeLists.txt]: see 3874966593e991cdeea9807de313311dccac52aa

File: include/antler/project/net_utils.hpp
Comment (ScottBailey) [reply include/antler/project/net_utils.hpp:1]: Not worth the hassle right now.

File: include/antler/project/dependency.hpp (hunk 2 above)
Comment (ScottBailey) [reply include/antler/project/dependency.hpp]: removed in 213c572c5b06f7e87c4cacacfd40e21c93188a0a

File: include/antler/project/dependency.hpp (hunk 3 above)
Comment (ScottBailey) [reply include/antler/project/dependency.hpp]: removed in 213c572c5b06f7e87c4cacacfd40e21c93188a0a

File: include/antler/project/dependency.hpp (hunk 4 above)
Comment (ScottBailey) [reply include/antler/project/dependency.hpp:83]: Good point, I agree, but probably not worth doing right now.

File: include/antler/project/mustache.hpp (hunk 5 above)
Comment (ScottBailey) [reply include/antler/project/mustache.hpp:42]: Leaving for now, might be used in next revision?

File: include/antler/project/location.hpp (hunk 18 above)
Comment (ScottBailey) [reply include/antler/project/location.hpp]: Further correction in 03260ee45452415776f21f6d6eb25ecd69704139

File: include/antler/project/project.hpp (hunk 6 above)
Comment (ScottBailey) [reply include/antler/project/project.hpp:91]: Added comment in 866aa28

File: include/antler/project/net_utils.hpp (hunk 21 above)
Comment (ScottBailey) [reply include/antler/project/net_utils.hpp:154]: Fixed in 03260ee4

File: tests/location_tests.cpp (hunk 7 above)
Comment (ScottBailey) [reply tests/location_tests.cpp:13]: This is passing for me. I'm not sure what, if anything, changed.

File: tests/location_tests.cpp (hunk 8 above)
Comment (ScottBailey) [reply tests/location_tests.cpp]: Re-enabled in 22e0e0f

File: tests/string_from.cpp (hunk 9 above)
Comment (ScottBailey) [reply tests/string_from.cpp]: Fixed in ab2fd78

File: tools/CMakeLists.txt (hunk 10 above)
Comment (ScottBailey) [reply tools/CMakeLists.txt:16]: Added TODO comment in 019fcda

File: tools/remove_from.hpp (hunk 14 above)
Comment (ScottBailey) [reply tools/remove_from.hpp]: Fixed in c1f01273

File: tools/tests/init_tests.py (hunk 15 above)
Comment (ScottBailey) [reply tools/tests/init_tests.py]: This test was fixed in f17b961

File: include/antler/project/populator.hpp (hunk 22 above)
Comment (ScottBailey) [reply include/antler/project/populator.hpp]: fixed in 6ce0391

File: include/antler/system/utils.hpp (hunk 26 above)
Comment (ScottBailey) [reply include/antler/system/utils.hpp]: Updated in 09a7323

File: include/antler/system/utils.hpp (hunk 28 above)
Comment (ScottBailey) [reply include/antler/system/utils.hpp]: fixed in 17da588

File: include/antler/system/utils.hpp (hunk 29 above)
Comment (ScottBailey) [reply include/antler/system/utils.hpp]: fixed in 17da588

File: licenses/LICENSE.Mustache
Comment (ScottBailey) [reply licenses/LICENSE.Mustache:1]: Actually this is a Boost license in the version we are using. Fixed in 1e8097e

File: src/dependency.cpp (hunk 30 above)
Comment (ScottBailey) [reply src/dependency.cpp]: fixed in 5b44f00

File: tools/remove_from.hpp (hunk 32 above)
Comment (ScottBailey) [reply tools/remove_from.hpp]: Unimportant, ignoring.

File: src/location.cpp (hunk 31 above)
Comment (ScottBailey) [reply src/location.cpp]: fixed in 8c9b5c7

File: tests/init_tests.cpp
Comment (ScottBailey) [reply tests/init_tests.cpp:1]: Fixed in f17b961b

**Comment by ScottBailey** [comment]
> The following files have "wrong" line endings:

Unimportant for now.

File: tests/location_tests.cpp (hunk 7 above)
Comment (dimas1185) [reply tests/location_tests.cpp]: I tried to play with it, so if I run it standalone - it passes, if I run it in parallel, it fails. It seems that it used same directories as other tests so I fixed it by renaming test project directory

File: CMakeLists.txt (hunk 16 above)
Comment (dimas1185) [reply CMakeLists.txt]: `CMAKE_CURRENT_BINARY_DIR` is something I added to allow it to be submoduled. I'm not sure about outcome of your suggested change but the goal is to have it in a current configured binary directory

File: licenses/LICENSE.cturtle [hunk 33]
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2019 - 2022 Daniil Goncharov
```
Comment (dimas1185) [inline licenses/LICENSE.cturtle:3]: why did you rename an author? Also, Bucky is not his legal name

File: CMakeLists.txt [hunk 34]
```diff
@@ -32,9 +32,9 @@ endif()
 #
 #  NOTE: This is used due to legacy requirements instead of:
 #    `install(FILE <license files> COMPONENT Documentation DESTINATION share/aproj/licenses)`
-configure_file(LICENSE ${CMAKE_CURRENT_BINARY_DIR}/licenses/LICENSE COPYONLY)
+configure_file(LICENSE licenses/LICENSE COPYONLY)
```
Comment (dimas1185) [inline CMakeLists.txt:34]: I may change this back in [cdt/#112](https://github.com/AntelopeIO/cdt/pull/114) if submoduling won't work with this

**Review by dimas1185 — APPROVED** [review]

File: CMakeLists.txt (hunk 34 above)
Comment (ScottBailey) [reply CMakeLists.txt:34]: Hmm, you might need to change it to something different. The replacement line and original should be functionally the same.

**Review by ScottBailey — APPROVED** [review]
